// ms/index.js
var original_module_exports = null;
var original_module_exports = null;
var original_module_exports = null;
var original_module_exports = null;
var original_module_exports = null;
var original_module_exports = null;
var original_module_exports = null;
var original_module_exports = null;
var original_plural = null;
var original_plural = null;
var original_plural = null;
var original_plural = null;
var original_plural = null;
var original_plural = null;
var original_plural = null;
var original_fmtLong = null;
var original_fmtLong = null;
var original_fmtLong = null;
var original_fmtLong = null;
var original_fmtLong = null;
var original_fmtLong = null;
var original_fmtLong = null;
var original_fmtShort = null;
var original_fmtShort = null;
var original_fmtShort = null;
var original_fmtShort = null;
var original_fmtShort = null;
var original_fmtShort = null;
var original_fmtShort = null;
var original_parse = null;
var original_parse = null;
var original_parse = null;
var original_parse = null;
var original_parse = null;
var original_parse = null;
var original_parse = null;
var srcFile = 'node_modules/ms/index.js';
var fs = require('fs');
var esprima = require('esprima');
var estraverse = require('estraverse');
var escodegen = require('escodegen');
var cachedCode = {};
var s = 1000;
var m = s * 60;
var h = m * 60;
var d = h * 24;
var y = d * 365.25;
module.exports = function (val, options) {
        if (original_module.exports == null) {
        lazyLoad(module.exports, srcFile);
        var loadedBody = loadAndInvoke('module.exports', srcFile);
        eval('original_module.exports = ' + loadedBody);
        module.exports = original_module.exports;
    }
        original_module.exports.apply(this, val, options);
};
{
    var original_parse;
    function  parse(str) {
                if (original_parse == null) {
            lazyLoad('parse', srcFile);
            var loadedBody = loadAndInvoke('parse', srcFile);
            eval('original_parse = ' + loadedBody);
            parse = original_parse;
        }
                original_parse.apply(this, str);
    }
}
{
    var original_fmtShort;
    function  fmtShort(ms) {
                if (original_fmtShort == null) {
            lazyLoad('fmtShort', srcFile);
            var loadedBody = loadAndInvoke('fmtShort', srcFile);
            eval('original_fmtShort = ' + loadedBody);
            fmtShort = original_fmtShort;
        }
                original_fmtShort.apply(this, ms);
    }
}
{
    var original_fmtLong;
    function  fmtLong(ms) {
                if (original_fmtLong == null) {
            lazyLoad('fmtLong', srcFile);
            var loadedBody = loadAndInvoke('fmtLong', srcFile);
            eval('original_fmtLong = ' + loadedBody);
            fmtLong = original_fmtLong;
        }
                original_fmtLong.apply(this, ms);
    }
}
{
    var original_plural;
    function  plural(ms, n, name) {
                if (original_plural == null) {
            lazyLoad('plural', srcFile);
            var loadedBody = loadAndInvoke('plural', srcFile);
            eval('original_plural = ' + loadedBody);
            plural = original_plural;
        }
                original_plural.apply(this, ms, n, name);
    }
}
function lazyLoad(funName, fileName) {
    var code = fs.readFileSync(fileName, 'utf8');
    var ast = esprima.parse(code.toString(), {
        range: true,
        loc: true,
        tokens: false
    });
    cachedCode[srcFile] = {};
    extractBodies(fileName);
}
function extractBodies(srcFile) {
    var code = fs.readFileSync(srcFile, 'utf8');
    var ast = esprima.parse(code.toString(), {
        range: true,
        loc: true,
        tokens: false
    });
    cachedCode[srcFile] = {};
    estraverse.traverse(ast, {
        enter: function (node, parent) {
            if (node.type == 'FunctionDeclaration') {
                var functionName = node.id.name;
                var functionBody = escodegen.generate(node);
                cachedCode[srcFile][functionName] = functionBody;
            } else if (node.type == 'ExpressionStatement') {
                if (node.expression.type == 'AssignmentExpression') {
                    var left = node.expression.left;
                    var right = node.expression.right;
                    if (right.type == 'FunctionExpression') {
                        if (left.type == 'MemberExpression') {
                            var leftVarBaseName = left.object.name;
                            var leftVarExtName = left.property.name;
                            var leftVarPath = leftVarBaseName + '.' + leftVarExtName;
                            var functionName = leftVarPath;
                            var functionBody = escodegen.generate(right);
                            cachedCode[srcFile][functionName] = functionBody;
                        }
                    }
                } else {
                    estraverse.VisitorOption.skip;
                }
            } else if (node.type == 'FunctionExpression') {
                var funName = null;
                if (node.id !== null) {
                    funName = node.id.name;
                } else {
                    funName = '_' + node.loc.start.line + '_' + node.loc.start.column;
                }
                if (funName !== null) {
                    cachedCode[srcFile][funName] = escodegen.generate(node);
                }
            } else {
                estraverse.VisitorOption.skip;
            }
        }
    });
}
function loadAndInvoke(funName, srcFile) {
    for (var elem in cachedCode) {
        if (cachedCode.hasOwnProperty(elem)) {
            if (elem === srcFile) {
                var functions = cachedCode[elem];
                for (var fun in functions) {
                    if (functions.hasOwnProperty(fun)) {
                        if (fun === funName) {
                            return functions[fun];
                        }
                    }
                }
            }
        }
    }
}
