// eslint-plugin-import/lib/rules/no-default-export.js
'use strict';

module.exports = {
  meta: {
    docs: {}
  },

  create: function create(context) {
    // ignore non-modules
    if (context.parserOptions.sourceType !== 'module') {
      return {};
    }

    var preferNamed = 'Prefer named exports.';
    var noAliasDefault = function noAliasDefault(_ref) {
      var local = _ref.local;
      return 'Do not alias `' + local.name + '` as `default`. Just export ' + ('`' + local.name + '` itself instead.');
    };

    return {
      ExportDefaultDeclaration: function ExportDefaultDeclaration(node) {
        context.report({ node: node, message: preferNamed });
      },
      ExportNamedDeclaration: function ExportNamedDeclaration(node) {
        node.specifiers.forEach(function (specifier) {
          if (specifier.type === 'ExportDefaultSpecifier' && specifier.exported.name === 'default') {
            context.report({ node: node, message: preferNamed });
          } else if (specifier.type === 'ExportSpecifier' && specifier.exported.name === 'default') {
            context.report({ node: node, message: noAliasDefault(specifier) });
          }
        });
      }
    };
  }
};
