// eslint-plugin-import/lib/rules/no-useless-path-segments.js
'use strict';

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _sumBy = require('lodash/sumBy');

var _sumBy2 = _interopRequireDefault(_sumBy);

var _resolve = require('eslint-module-utils/resolve');

var _resolve2 = _interopRequireDefault(_resolve);

var _moduleVisitor = require('eslint-module-utils/moduleVisitor');

var _moduleVisitor2 = _interopRequireDefault(_moduleVisitor);

var _docsUrl = require('../docsUrl');

var _docsUrl2 = _interopRequireDefault(_docsUrl);

function _interopRequireDefault(obj) {
  return obj && obj.__esModule ? obj : { default: obj };
}

/**
 * convert a potentially relative path from node utils into a true
 * relative path.
 *
 * ../ -> ..
 * ./ -> .
 * .foo/bar -> ./.foo/bar
 * ..foo/bar -> ./..foo/bar
 * foo/bar -> ./foo/bar
 *
 * @param rel {string} relative posix path potentially missing leading './'
 * @returns {string} relative posix path that always starts with a ./
 **/
function toRel(rel) {
  var stripped = rel.replace(/\/$/g, '');
  return (/^((\.\.)|(\.))($|\/)/.test(stripped) ? stripped : './' + stripped
  );
} /**
   * @fileOverview Ensures that there are no useless path segments
   * @author Thomas Grainger
   */

function normalize(fn) {
  return toRel(_path2.default.posix.normalize(fn));
}

var countRelParent = function countRelParent(x) {
  return (0, _sumBy2.default)(x, function (v) {
    return v === '..';
  });
};

module.exports = {
  meta: {
    docs: {
      url: (0, _docsUrl2.default)('no-useless-path-segments')
    },

    fixable: 'code'
  },

  create: function create(context) {
    var currentDir = _path2.default.dirname(context.getFilename());

    function checkSourceValue(source) {
      var value = source.value;

      function report(proposed) {
        context.report({
          node: source,
          message: 'Useless path segments for "' + value + '", should be "' + proposed + '"',
          fix: function fix(fixer) {
            return fixer.replaceText(source, JSON.stringify(proposed));
          }
        });
      }

      if (!value.startsWith('.')) {
        return;
      }

      var resolvedPath = (0, _resolve2.default)(value, context);
      var normed = normalize(value);
      if (normed !== value && resolvedPath === (0, _resolve2.default)(normed, context)) {
        return report(normed);
      }

      if (value.startsWith('./')) {
        return;
      }

      if (resolvedPath === undefined) {
        return;
      }

      var expected = _path2.default.relative(currentDir, resolvedPath);
      var expectedSplit = expected.split(_path2.default.sep);
      var valueSplit = value.replace(/^\.\//, '').split('/');
      var valueNRelParents = countRelParent(valueSplit);
      var expectedNRelParents = countRelParent(expectedSplit);
      var diff = valueNRelParents - expectedNRelParents;

      if (diff <= 0) {
        return;
      }

      return report(toRel(valueSplit.slice(0, expectedNRelParents).concat(valueSplit.slice(valueNRelParents + diff)).join('/')));
    }

    return (0, _moduleVisitor2.default)(checkSourceValue, context.options[0]);
  }
};
